File: app/src/main/java/me/devsaki/hentoid/parsers/images/AnchiraParser.java
package me.devsaki.hentoid.parsers.images;

import android.os.Handler;
import android.os.Looper;
import android.webkit.URLUtil;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import me.devsaki.hentoid.activities.sources.WebResultConsumer;
import me.devsaki.hentoid.core.HentoidApp;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.enums.Site;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.parsers.ParseHelper;
import me.devsaki.hentoid.util.Helper;
import me.devsaki.hentoid.util.StringHelper;
import me.devsaki.hentoid.util.exception.EmptyResultException;
import me.devsaki.hentoid.util.exception.ParseException;
import me.devsaki.hentoid.util.network.HttpHelper;
import me.devsaki.hentoid.views.AnchiraBackgroundWebView;
import timber.log.Timber;

/**
 * Handles parsing of content from anchira.to
 */
public class AnchiraParser extends BaseImageListParser implements WebResultConsumer {

    private final AtomicInteger resultCode = new AtomicInteger(-1);
    private final AtomicReference<Content> resultContent = new AtomicReference<>();
    private AnchiraBackgroundWebView anchiraWv = null;

    @Override
    public List<ImageFile> parseImageListImpl(@NonNull Content onlineContent, @Nullable Content storedContent) throws Exception {
        String readerUrl = onlineContent.getReaderUrl();
        processedUrl = onlineContent.getGalleryUrl();

        if (!URLUtil.isValidUrl(readerUrl))
            throw new IllegalArgumentException("Invalid gallery URL : " + readerUrl);

        Timber.d("Gallery URL: %s", readerUrl);

        EventBus.getDefault().register(this);

        List<ImageFile> result;
        try {
            result = parseImageListWithWebview(onlineContent);
            ParseHelper.setDownloadParams(result, onlineContent.getSite().getUrl());
        } catch (Exception e) {
            Helper.logException(e);
            result = new ArrayList<>();
        } finally {
            EventBus.getDefault().unregister(this);
        }

        return result;
    }

    public Content parseContentWithWebview(@NonNull String url) throws Exception {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(() -> {
            anchiraWv = new AnchiraBackgroundWebView(HentoidApp.Companion.getInstance(), this, Site.ANCHIRA);
            Timber.d(">> loading url %s", url);
            anchiraWv.loadUrl(url);
            Timber.i(">> loading wv");
        });

        var remainingIterations = 30; // Timeout
        while (-1 == resultCode.get() && remainingIterations-- > 0 && !processHalted.get())
            Helper.pause(500);

        if (processHalted.get())
            throw new EmptyResultException("Unable to detect content (empty result)");

        synchronized (resultCode) {
            int res = resultCode.get();
            if (0 == res) {
                Content c = resultContent.get();
                if (c != null) return c;
            } else if (-1 == res) {
                throw new ParseException("Parsing failed to start");
            } else if (2 == res) {
                throw new ParseException("Parsing has failed unexpectedly");
            }
            throw new EmptyResultException("Parsing hasn't found any content");
        }
    }

    public List<ImageFile> parseImageListWithWebview(@NonNull Content onlineContent) throws Exception {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(() -> {
            anchiraWv = new AnchiraBackgroundWebView(HentoidApp.Companion.getInstance(), this, Site.ANCHIRA);
            String pageUrl = onlineContent.getReaderUrl();
            Timber.d(">> loading url %s", pageUrl);
            anchiraWv.loadUrl(pageUrl);
            Timber.i(">> loading wv");
        });

        var remainingIterations = 30; // Timeout
        while (-1 == resultCode.get() && remainingIterations-- > 0 && !processHalted.get())
            Helper.pause(500);

        if (processHalted.get())
            throw new EmptyResultException("Unable to detect pages (empty result)");

        synchronized (resultCode) {
            int res = resultCode.get();
            if (0 == res) {
                Content c = resultContent.get();
                if (c != null) {
                    String imgUrl = c.getCoverImageUrl();
                    HttpHelper.UriParts parts = new HttpHelper.UriParts(imgUrl);
                    String fileName = parts.getFileNameNoExt();
                    int length = fileName.length();

                    List<String> urls = new ArrayList<>();
                    for (int i = 1; i <= onlineContent.getQtyPages(); i++) {
                        parts.setFileNameNoExt(StringHelper.formatIntAsStr(i, length));
                        parts.setExtension((1 == i) ? "jpg" : "png"); // Try to minimize failed requests
                        urls.add(parts.toUri());
                    }

                    return ParseHelper.urlsToImageFiles(urls, onlineContent.getCoverImageUrl(), StatusContent.SAVED);
                }
            } else if (-1 == res) {
                throw new ParseException("Parsing failed to start");
            } else if (2 == res) {
                throw new ParseException("Parsing has failed unexpectedly");
            }
            throw new EmptyResultException("Parsing hasn't found any page");
        }
    }

    public void destroy() {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(() -> {
            if (anchiraWv != null) anchiraWv.destroy();
            anchiraWv = null;
        });
    }

    @Override
    public String getAltUrl(@NonNull String url) {
        HttpHelper.UriParts parts = new HttpHelper.UriParts(url);
        String ext = parts.getExtension();
        String altExt = (ext.equalsIgnoreCase("jpg")) ? "png" : "jpg";
        parts.setExtension(altExt);
        return parts.toUri();
    }

    @Override
    protected List<String> parseImages(@NonNull Content content) {
        // We won't use that as parseImageListImpl is overriden directly
        return null;
    }

    @Override
    public void onContentReady(@NonNull Content result, boolean quickDownload) {
        resultContent.set(result);
        resultCode.set(0);
    }

    @Override
    public void onNoResult() {
        resultCode.set(1);
    }

    @Override
    public void onResultFailed() {
        resultCode.set(2);
    }
}

File: app/src/main/java/me/devsaki/hentoid/parsers/images/AnchiraParser.kt
package me.devsaki.hentoid.parsers.images

import android.os.Handler
import android.os.Looper
import android.webkit.URLUtil
import androidx.core.util.Pair
import me.devsaki.hentoid.activities.sources.WebResultConsumer
import me.devsaki.hentoid.core.HentoidApp.Companion.getInstance
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.parsers.ParseHelper
import me.devsaki.hentoid.util.Helper
import me.devsaki.hentoid.util.StringHelper
import me.devsaki.hentoid.util.exception.EmptyResultException
import me.devsaki.hentoid.util.exception.ParseException
import me.devsaki.hentoid.util.network.HttpHelper.UriParts
import me.devsaki.hentoid.views.AnchiraBackgroundWebView
import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

class AnchiraParser : BaseImageListParser(), WebResultConsumer {
    private val resultCode = AtomicInteger(-1)
    private val resultContent = AtomicReference<Content>()
    private var anchiraWv: AnchiraBackgroundWebView? = null

    @Throws(Exception::class)
    override fun parseImageListImpl(
        onlineContent: Content,
        storedContent: Content?
    ): List<ImageFile> {
        val readerUrl = onlineContent.readerUrl
        processedUrl = onlineContent.galleryUrl
        require(URLUtil.isValidUrl(readerUrl)) { "Invalid gallery URL : $readerUrl" }
        Timber.d("Gallery URL: %s", readerUrl)
        EventBus.getDefault().register(this)
        var result: List<ImageFile>
        try {
            result = parseImageListWithWebview(onlineContent)
            ParseHelper.setDownloadParams(result, onlineContent.site.url)
        } catch (e: Exception) {
            Helper.logException(e)
            result = emptyList()
        } finally {
            EventBus.getDefault().unregister(this)
            clear()
        }
        return result
    }

    @Throws(Exception::class)
    fun parseContentWithWebview(url: String): Content {
        val handler = Handler(Looper.getMainLooper())
        handler.post {
            anchiraWv = AnchiraBackgroundWebView(
                getInstance(),
                this,
                Site.ANCHIRA,
                0
            )
            Timber.d(">> loading url %s", url)
            anchiraWv?.loadUrl(url)
            Timber.i(">> loading wv")
        }

        var remainingIterations = 30 // Timeout
        while (-1 == resultCode.get() && remainingIterations-- > 0 && !processHalted.get())
            Helper.pause(500)
        if (processHalted.get()) throw EmptyResultException("Unable to detect content (empty result)")

        synchronized(resultCode) {
            val res = resultCode.get()
            if (0 == res) {
                val c = resultContent.get()
                if (c != null) return c
            } else if (-1 == res) {
                throw ParseException("Parsing failed to start")
            } else if (2 == res) {
                throw ParseException("Parsing has failed unexpectedly")
            }
            throw EmptyResultException("Parsing hasn't found any content")
        }
    }

    @Throws(Exception::class)
    fun parseImageListWithWebview(onlineContent: Content): List<ImageFile> {
        val handler = Handler(Looper.getMainLooper())
        handler.post {
            anchiraWv = AnchiraBackgroundWebView(
                getInstance(),
                this,
                Site.ANCHIRA,
                1
            )
            val pageUrl = onlineContent.galleryUrl
            Timber.d(">> loading url %s", pageUrl)
            anchiraWv?.loadUrl(pageUrl)
            Timber.i(">> loading wv")
        }

        var remainingIterations = 30 // Timeout
        while (-1 == resultCode.get() && remainingIterations-- > 0 && !processHalted.get())
            Helper.pause(500)
        if (processHalted.get()) throw EmptyResultException("Unable to detect pages (empty result)")

        synchronized(resultCode) {
            val res = resultCode.get()
            if (0 == res) {
                val c = resultContent.get()
                // Might fail when called from a merged book where qtyPages accounts for all chapters
                val nbPages = onlineContent.qtyPages

                if (c != null) {
                    val imgUrl = c.coverImageUrl
                    Timber.d(">> retrieving url %s", imgUrl)
                    val parts = UriParts(imgUrl, false)
                    val fileName = parts.fileNameNoExt
                    val urls: MutableList<String> =
                        ArrayList()
                    if (StringHelper.isNumeric(fileName)) {
                        val length = fileName.length
                        for (i in 1..nbPages) {
                            parts.fileNameNoExt = StringHelper.formatIntAsStr(i, length)
                            parts.extension =
                                if (1 == i) "jpg" else "png" // Try to minimize failed requests
                            urls.add(parts.toUri())
                        }
                    } else {
                        val startIndex = fileName.lastIndexOf("-%20") + 4
                        val endIndex = fileName.lastIndexOf("%20(")
                        val length = endIndex - startIndex
                        val part1 = fileName.substring(0, startIndex)
                        val part2 = fileName.substring(endIndex)
                        for (i in 1..nbPages) {
                            parts.fileNameNoExt = part1 + StringHelper.formatIntAsStr(
                                i,
                                length
                            ) + part2
                            parts.extension =
                                if (1 == i) "jpg" else "png" // Try to minimize failed requests
                            urls.add(parts.toUri())
                        }
                    }
                    return ParseHelper.urlsToImageFiles(
                        urls,
                        onlineContent.coverImageUrl,
                        StatusContent.SAVED
                    )
                }
            } else if (-1 == res) {
                throw ParseException("Parsing failed to start")
            } else if (2 == res) {
                throw ParseException("Parsing has failed unexpectedly")
            }
            throw EmptyResultException("Parsing hasn't found any page")
        }
    }

    override fun clear() {
        val handler = Handler(Looper.getMainLooper())
        handler.post {
            anchiraWv?.destroy()
            anchiraWv = null
        }
    }

    override fun getAltUrl(url: String): String {
        val parts = UriParts(url, false)
        val ext = parts.extension
        val altExt = if (ext.equals("jpg", ignoreCase = true)) "png" else "jpg"
        parts.extension = altExt
        return parts.toUri()
    }

    override fun isChapterUrl(url: String): Boolean {
        // No chapters for this source
        return false
    }

    override fun parseImages(content: Content): List<String> {
        // We won't use that as parseImageListImpl is overriden directly
        return emptyList()
    }

    override fun parseImages(
        chapterUrl: String,
        downloadParams: String?,
        headers: List<Pair<String, String>>?
    ): List<String> {
        // We won't use that as parseImageListImpl is overriden directly
        return emptyList()
    }

    override fun onContentReady(result: Content, quickDownload: Boolean) {
        resultContent.set(result)
        resultCode.set(0)
    }

    override fun onNoResult() {
        resultCode.set(1)
    }

    override fun onResultFailed() {
        resultCode.set(2)
    }
}